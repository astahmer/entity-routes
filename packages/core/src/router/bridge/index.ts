export * from "./BridgeRouter";
export * from "./ContextAdapter";
export * from "./express";
