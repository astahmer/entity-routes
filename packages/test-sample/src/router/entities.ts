import { IsString } from "class-validator";
import { Column, Entity, ManyToOne, OneToMany, PrimaryGeneratedColumn } from "typeorm";

import { EntityRoute, Groups, Subresource } from "@entity-routes/core";

export const getTestEntities = () => [User, Article, Comment, Upvote, Image, Role];

class AbstractEntity {
    @Groups("all")
    @PrimaryGeneratedColumn()
    id: number;
}

@EntityRoute()
@Entity()
class Image extends AbstractEntity {
    @Groups("all")
    @Column()
    url: string;

    @Subresource(() => Upvote)
    @OneToMany(() => Upvote, (upvote) => upvote.image)
    upvotes: () => Upvote[]; // wrap in fn to avoid ReferenceError: Cannot access 'Upvote' before initialization

    @OneToMany(() => Role, (role) => role.logo)
    logoOfRoles: () => Role; // wrap in fn to avoid ReferenceError: Cannot access 'Role' before initialization
}

@EntityRoute()
@Entity()
class Role extends AbstractEntity {
    @Groups("all")
    @Column()
    label: string;

    @Subresource(() => Image)
    @ManyToOne(() => Image, (image) => image.logoOfRoles)
    logo: Image;

    @OneToMany(() => User, (user) => user.mainRole)
    mainRoleOfUsers: () => User;
}

@EntityRoute({ operations: ["create", "update", "details", "list", "delete"] })
@Entity()
class User extends AbstractEntity {
    @Groups("all")
    @IsString()
    @Column()
    name: string;

    @Subresource(() => Article, { maxDepth: 3 })
    @OneToMany(() => Article, (article) => article.author)
    articles: Article[];

    @Subresource(() => Role, { maxDepth: 3 })
    @ManyToOne(() => Role, (role) => role.mainRoleOfUsers)
    mainRole: Role;
}

@EntityRoute({ operations: ["details"] })
@Entity()
class Article extends AbstractEntity {
    @Groups({ article: ["create"] })
    @Column()
    title: string;

    @Subresource(() => User)
    @ManyToOne(() => User, (user) => user.articles)
    author: User;

    @Subresource(() => Comment)
    @OneToMany(() => Comment, (comment) => comment.article)
    comments: Comment[];
}

@EntityRoute()
@Entity()
class Comment extends AbstractEntity {
    @Groups({ comment: ["create"] })
    @Column()
    message: string;

    @ManyToOne(() => Article, (article) => article.comments)
    article: Article;

    @Subresource(() => Upvote)
    @OneToMany(() => Upvote, (upvote) => upvote.comment)
    upvotes: Upvote[];
}

@EntityRoute()
@Entity()
class Upvote extends AbstractEntity {
    @ManyToOne(() => Comment, (comment) => comment.upvotes)
    comment: Comment;

    @ManyToOne(() => Image, (image) => image.upvotes)
    image: Image;
}
