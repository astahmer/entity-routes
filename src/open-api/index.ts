export * from "./baseReferences";
export * from "./helpers";
export * from "./maker";
