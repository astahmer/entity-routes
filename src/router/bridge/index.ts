export * from "./BridgeRouter";
export * from "./ContextAdapter";
export * from "./express";
export * from "./koa";
