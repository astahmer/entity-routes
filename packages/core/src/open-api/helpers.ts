import { SchemaObject } from "openapi3-ts";
import { ColumnType } from "typeorm";

import { upperFirstLetter } from "@entity-routes/shared";

import { RouteOperation } from "../decorators";
import { RouteResponseType } from "../router";

const idParamType = "(\\d+)";

/** Replace all :entityId(\\d+) by {entityId} */
export const getOpenApiPath = (path: string) =>
    path.split(idParamType).join("").split(":").join("{").split("{id").join("{id}").split("d/").join("d}/");

export const routeResponseTypeByOperation: Record<RouteOperation, RouteResponseType> = {
    create: "persist",
    update: "persist",
    list: "collection",
    details: "item",
    delete: "delete",
};
export const getBaseRouteResponseSchemaByType = (type: RouteResponseType) =>
    `Base${upperFirstLetter(type)}RouteResponseSchema`;

export type OpenApiPrimitive = Extract<SchemaObject["type"], "integer" | "number" | "string" | "boolean">;
/**
 * Associations of TypeORM.ColumnType with OpenAPI.DataType
 * @see https://github.com/typeorm/typeorm/blob/master/docs/entities.md#column-types-for-mysql--mariadb
 * @see https://github.com/typeorm/typeorm/blob/master/src/driver/types/ColumnTypes.ts
 */
export const openApiSchemaTypeByColumnType: Partial<Record<Extract<ColumnType, string>, OpenApiPrimitive>> = {
    //
    number: "number",
    float: "number",
    //
    int: "integer",
    integer: "integer",
    //
    bool: "boolean",
    boolean: "boolean",
    //
    date: "string",
    datetime: "string",
    time: "string",
    timestamp: "string",
    //
    varchar: "string",
    text: "string",
    string: "string",
};
/** Get OpenAPI SchemObject.type by TypeORM.ColumnType, if no direct association -> defaults to string */
export const typeOrmTypeToOpenApiType = (type: ColumnType) =>
    (typeof type === "string"
        ? openApiSchemaTypeByColumnType[type]
        : openApiSchemaTypeByColumnType[type.name as keyof typeof openApiSchemaTypeByColumnType]) || "string";

export const makeResponseFromSchema = (schema: SchemaObject, description = "OK") => ({
    description,
    content: { "application/json": { schema } },
});
export const makeResponseWithRef = (ref: string, description: string = "OK") =>
    makeResponseFromSchema({ $ref: `#/components/schemas/${ref}` }, description);
