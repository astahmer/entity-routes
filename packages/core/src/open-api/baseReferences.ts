import { OpenApiBuilder } from "openapi3-ts";

import { makeResponseWithRef } from "./helpers";

/** Add ts types as OpenAPI references (responses, schema, etc) */
export function addComponentsReferences(builder: OpenApiBuilder) {
    // Responses using schemas defined below
    builder.addResponse("BaseRouteResponse", makeResponseWithRef("BaseRouteResponseSchema"));
    builder.addResponse("ErrorRouteResponse", makeResponseWithRef("BaseErrorRouteResponseSchema", "Response error."));

    // Parameters
    builder.addParameter("EntityId", {
        name: "id",
        in: "path",
        required: true,
        schema: { type: "integer" },
        description: "Current entity route id",
    });
    // TODO PaginationFilter/OrderFilter
    // TODO Mapping.pretty

    // Schemas
    // @context
    builder.addSchema("BaseRouteContext", {
        type: "object",
        required: ["entity", "operation"],
        properties: { entity: { type: "string" }, operation: { type: "string" } },
    });
    builder.addSchema("CollectionRouteContext", {
        allOf: [{ $ref: "#/components/schemas/BaseRouteContext" }],
        type: "object",
        required: ["retrievedItems"],
        properties: { totalItems: { type: "integer" }, retrievedItems: { type: "integer" } },
    });
    builder.addSchema("PersistRouteContext", {
        allOf: [{ $ref: "#/components/schemas/BaseRouteContext" }],
        type: "object",
        required: ["validationErrors"],
        properties: { validationErrors: { $ref: "#/components/schemas/EntityErrorResults" } },
    });
    builder.addSchema("ErrorRouteContext", {
        allOf: [{ $ref: "#/components/schemas/BaseRouteContext" }],
        type: "object",
        required: ["error"],
        properties: { error: { type: "string" } },
    });
    // @context.validationErrors
    builder.addSchema("EntityErrorResults", {
        type: "object",
        additionalProperties: { type: "array", items: { $ref: "#/components/schemas/EntityError" } },
    });
    builder.addSchema("EntityError", {
        type: "object",
        properties: {
            currentPath: { type: "string" },
            property: { type: "string" },
            constraints: { additionalProperties: { type: "string" } },
        },
    });
    // RouteResponse
    builder.addSchema("BaseRouteResponseSchema", {
        type: "object",
        required: ["@context"],
        properties: { "@context": { $ref: "#/components/schemas/BaseRouteContext" } },
    });
    builder.addSchema("BaseItemRouteResponseSchema", {
        type: "object",
        allOf: [{ $ref: "#/components/schemas/BaseRouteResponseSchema" }],
    });
    builder.addSchema("BaseCollectionRouteResponseSchema", {
        type: "object",
        properties: { "@context": { $ref: "#/components/schemas/CollectionRouteContext" } },
    });
    builder.addSchema("BasePersistRouteResponseSchema", {
        type: "object",
        properties: { "@context": { $ref: "#/components/schemas/PersistRouteContext" } },
    });
    builder.addSchema("BaseErrorRouteResponseSchema", {
        type: "object",
        properties: { "@context": { $ref: "#/components/schemas/ErrorRouteContext" } },
    });
    builder.addSchema("BaseDeleteRouteResponseSchema", {
        type: "object",
        allOf: [{ $ref: "#/components/schemas/BaseRouteResponseSchema" }],
        properties: { deleted: { title: "TypeORM.DeleteResult" } },
    });

    return builder;
}
