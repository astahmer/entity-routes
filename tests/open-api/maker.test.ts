import { Server } from "net";
import { makeOpenApiBuilderFrom, makeOpenApi } from "@/open-api";
import { setupKoaApp } from "@@/tests/router/bridge/koaSetup";
import { closeTestConnection } from "@@/tests/testConnection";

import { getOpenApiTestEntities } from "../mapping/sample/entities";

describe("OpenAPI", () => {
    const entities = getOpenApiTestEntities();
    let server: Server;

    beforeAll(async () => {
        const result = await setupKoaApp(entities);
        server = result.server;
    });
    afterAll(() => {
        server.close();
        return closeTestConnection();
    });

    it("exportRoutesAsOpenApi", async () => {
        const openApi = makeOpenApi();
        const builder = makeOpenApiBuilderFrom(openApi);
        const spec = builder.getSpecAsYaml();
        expect(spec).toMatchSnapshot();
    });
});
