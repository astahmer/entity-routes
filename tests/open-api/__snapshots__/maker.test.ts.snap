// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI exportRoutesAsOpenApi 1`] = `
"openapi: 3.0.0
info:
  title: title
  version: 1.0.0
  description: entity-routes OpenAPI spec
paths:
  /role:
    post:
      description: role_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters: []
    get:
      description: role_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  /role/mapping:
    post:
      description: role_create_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters: []
    get:
      description: role_list_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/role/{id}\\":
    get:
      description: role_details
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    put:
      description: role_update
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    delete:
      description: role_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/role/{id}/mapping\\":
    get:
      description: role_details_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    put:
      description: role_update_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/role/{RoleId}/users\\":
    post:
      description: role_users_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/RoleId\\"
    get:
      description: role_users_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/role/{RoleId}/users/{id}\\":
    delete:
      description: role_users_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/RoleId\\"
      - $ref: \\"#/components/parameters/EntityId\\"
  \\"/role/{RoleId}/users/articles\\":
    get:
      description: role_users_articles_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      avatar:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
                      id:
                        type: string
                      name:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/RoleId\\"
  /user:
    post:
      description: user_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters: []
    get:
      description: user_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  /user/mapping:
    post:
      description: user_create_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters: []
    get:
      description: user_list_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/user/{id}\\":
    get:
      description: user_details
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
                  roles:
                    properties:
                      id:
                        type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    put:
      description: user_update
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    delete:
      description: user_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/user/{id}/mapping\\":
    get:
      description: user_details_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
                  roles:
                    properties:
                      id:
                        type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    put:
      description: user_update_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  avatar:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  name:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/user/{UserId}/articles\\":
    post:
      description: user_articles_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/UserId\\"
    get:
      description: user_articles_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      avatar:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
                      id:
                        type: string
                      name:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/user/{UserId}/articles/{id}\\":
    delete:
      description: user_articles_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/UserId\\"
      - $ref: \\"#/components/parameters/EntityId\\"
  \\"/user/{UserId}/articles/comments\\":
    get:
      description: user_articles_comments_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  message:
                    type: string
                  upvotes:
                    properties:
                      id:
                        type: string
                      reaction:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/UserId\\"
  \\"/user/{UserId}/roles\\":
    post:
      description: user_roles_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/UserId\\"
    get:
      description: user_roles_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  logo:
                    properties:
                      id:
                        type: string
                      url:
                        type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/user/{UserId}/roles/{id}\\":
    delete:
      description: user_roles_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/UserId\\"
      - $ref: \\"#/components/parameters/EntityId\\"
  /article:
    post:
      description: article_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters: []
    get:
      description: article_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      avatar:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
                      id:
                        type: string
                      name:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  /article/mapping:
    post:
      description: article_create_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      id:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters: []
    get:
      description: article_list_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      avatar:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
                      id:
                        type: string
                      name:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/article/{id}\\":
    get:
      description: article_details
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      avatar:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
                      id:
                        type: string
                      name:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    put:
      description: article_update
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    delete:
      description: article_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/article/{id}/mapping\\":
    get:
      description: article_details_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  author:
                    properties:
                      avatar:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
                      id:
                        type: string
                      name:
                        type: string
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    put:
      description: article_update_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  title:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/article/{ArticleId}/comments\\":
    post:
      description: article_comments_create
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  message:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/ArticleId\\"
    get:
      description: article_comments_list
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
                  message:
                    type: string
                  upvotes:
                    properties:
                      id:
                        type: string
                      reaction:
                        properties:
                          id:
                            type: string
                          url:
                            type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/article/{ArticleId}/comments/{id}\\":
    delete:
      description: article_comments_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/ArticleId\\"
      - $ref: \\"#/components/parameters/EntityId\\"
  \\"/comment/{id}\\":
    put:
      description: comment_update
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
    delete:
      description: comment_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  \\"/comment/{id}/mapping\\":
    put:
      description: comment_update_mapping
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  id:
                    type: string
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
  \\"/upvote/{id}\\":
    delete:
      description: upvote_delete
      responses:
        \\"200\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseDeleteRouteResponseSchema\\"
        \\"500\\":
          description: OK
          content:
            application/json:
              schema:
                $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
    parameters:
      - $ref: \\"#/components/parameters/EntityId\\"
components:
  schemas:
    BaseRouteContext:
      type: object
      required:
        - entity
        - operation
      properties:
        entity:
          type: string
        operation:
          type: string
    CollectionRouteContext:
      allOf:
        - $ref: \\"#/components/schemas/BaseRouteContext\\"
      type: object
      required:
        - retrievedItems
      properties:
        totalItems:
          type: integer
        retrievedItems:
          type: integer
    PersistRouteContext:
      allOf:
        - $ref: \\"#/components/schemas/BaseRouteContext\\"
      type: object
      required:
        - validationErrors
      properties:
        validationErrors:
          $ref: \\"#/components/schemas/EntityErrorResults\\"
    ErrorRouteContext:
      allOf:
        - $ref: \\"#/components/schemas/BaseRouteContext\\"
      type: object
      required:
        - error
      properties:
        error:
          type: string
    EntityErrorResults:
      type: object
      additionalProperties:
        type: array
        items:
          $ref: \\"#/components/schemas/EntityError\\"
    EntityError:
      type: object
      properties:
        currentPath:
          type: string
        property:
          type: string
        constraints:
          additionalProperties:
            type: string
    BaseRouteResponseSchema:
      type: object
      required:
        - \\"@context\\"
      properties:
        \\"@context\\":
          $ref: \\"#/components/schemas/BaseRouteContext\\"
    BaseItemRouteResponseSchema:
      type: object
      allOf:
        - $ref: \\"#/components/schemas/BaseRouteResponseSchema\\"
    BaseCollectionRouteResponseSchema:
      type: object
      properties:
        \\"@context\\":
          $ref: \\"#/components/schemas/CollectionRouteContext\\"
    BasePersistRouteResponseSchema:
      type: object
      properties:
        \\"@context\\":
          $ref: \\"#/components/schemas/PersistRouteContext\\"
    BaseErrorRouteResponseSchema:
      type: object
      properties:
        \\"@context\\":
          $ref: \\"#/components/schemas/ErrorRouteContext\\"
    BaseDeleteRouteResponseSchema:
      type: object
      allOf:
        - $ref: \\"#/components/schemas/BaseRouteResponseSchema\\"
      properties:
        deleted:
          title: TypeORM.DeleteResult
  responses:
    BaseRouteResponse:
      description: OK
      content:
        application/json:
          schema:
            $ref: \\"#/components/schemas/BaseRouteResponseSchema\\"
    ErrorRouteResponse:
      description: Response error.
      content:
        application/json:
          schema:
            $ref: \\"#/components/schemas/BaseErrorRouteResponseSchema\\"
  parameters:
    RoleId:
      name: RoleId
      in: path
      required: true
      schema:
        type: integer
    UserId:
      name: UserId
      in: path
      required: true
      schema:
        type: integer
    ArticleId:
      name: ArticleId
      in: path
      required: true
      schema:
        type: integer
    EntityId:
      name: id
      in: path
      required: true
      schema:
        type: integer
      description: Current entity route id
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers: []
"
`;
